examples/FlinkDeployment.yaml

```diff
@@ -18,7 +18,7 @@ spec:
       memory: "2048m"
       cpu: 1
   job:
-    jarURI: local:///opt/flink/usrlib/flink-sql-runner.jar
+    jarURI: local:///opt/flink/streamshub/flink-sql-runner-0.0.1-SNAPSHOT.jar
```

Comment:
that should be /opt/flink/opt/streamshub. I think putting the job jar in /opt/flink/usrlib might be a bit dodgy as I think it will be on the flink classpath as well as being submitted via the rest api.

also, maybe we should be creating a symlink with the version removed since users shouldn't have to update the version in two places (image and jar uri) when they upgrade.

Reply:
though I think it's correct for the distribution tarball to include it with version

Reply:
Why nest under `/opt/flink` why not `/opt/streamshub`?

>  I think it's correct for the distribution tarball to include it with version

Yes, that seems right, but what value does it add? The distribution is already versioned so what extra value would users get from having the version repeated again? I'm wondering if we should strip the version out when assembling the distribution.

Reply:
mm true, we could change the flink-amq-streams dist to contain `flink` and `streamshub` directories and unpack it into opt

Reply:
pushed that up, now the image contains:

```
/opt/flink # matches upstream flink distribution
/opt/streamshub/flink-sql-runner.jar # job jar
/opt/streamshub/usrlib/ # additional jars we want on the flink classpath (connectors/formats)
```

Reply:
@robobario does this issue concern us? https://issues.apache.org/jira/browse/FLINK-29288

Reply:
Thanks Tina, it may do, some things I've learned through that:

If you use [standalone](hhttps://nightlies.apache.org/flink/flink-kubernetes-operator-docs-release-1.9/docs/custom-resource/overview/#cluster-deployment-modes) mode (native is the default), then it volume mounts /opt/flink/usrlib so anything in the image is not available. I tried setting mode standalone and it makes the job manager blow up with `Caused by: java.nio.file.NoSuchFileException: /opt/flink/usrlib/flink-sql-runner.jar`

Then from further reading/experiments I've found that `/opt/flink/usrlib` only appears to be relevant to application mode. See the deployment [overview docs](https://nightlies.apache.org/flink/flink-docs-release-1.20/docs/deployment/overview/#application-mode). If we run in session mode it's not included in the classpath.

The ticket also shows we have the option (in application mode) of putting the job jar on the classpath, and not setting jarURI (I wonder if we would need to set an entry class, or if it would discover the main class via job jar manifest). Putting the job jar into /opt/flink/lib might cause issues in session mode similar to the ticket.

In summary:

1. installing connectors/formats into `usrlib` looks like it won't work with standalone deployment or session cluster (have tried it out and the jobmanager fails to start the job because it can't access the kafka connector)
2. installing connectors/formats into `/opt/flink/lib` would work for native/standalone and application/session.
3. shading connectors/formats into the job jar would work for native/standalone and application/session.

I think option 1 is ruled out since we should support session mode.

Option 2 changes the core flink classpath, so we have to be careful not to introduce any class collisions. It is more visible to tooling what jars are in the image, so we might get a warning if there is a vuln in the connector/format jar, less likely if it's shaded into a custom jar.

Option 3 is closest to how upstream users generally use flink, packing connectors into job jars. So we would be less likely to hit unique classloading problems. It's less visible what's included in the image so might be harder for static analysis to detect vulnerable jars in the image.

Maybe going with option 3 will cause the least heartache as we get started.

Reply:
> The distribution is already versioned so what extra value would users get from having the version repeated again? 

for SAST scanning purposes and our own sanity I think it's nice to see that the image contains sql-runner-x.y.z.jar (have currently removed the version so that it's called sql-runner.jar)

Reply:
Have changed it so the version is included in the dist and image, but we also create a symlink in the docker image so that the FlinkDeployment can refer to the stable `/opt/streamshub/flink-sql-runner.jar`

Reply:
In the end I implemented option 2 and checked it works with standalone/native mode and session/application modes.

Reply:
Thanks @robobario for looking into it. Reading the docs you have linked and the comments from the issue again, I got the same understanding as you summarised here. Going with option 2 looks good to me.